examples/microphone_nextjs/utils/recorder.ts

```diff
@@ -0,0 +1,217 @@
+    }
+    this.updateDeviceList();
+    return stream;
+  };
+
+  // getDevices is used to prompt the user to give permission to audio inputs
+  public getDevices = async () => {
+    // We first check if the system is busy - we don't want to prompt for permissions if the user is already prompted for permissions
+    if (!this.busy) {
+      this.busy = true;
+      await this.promptAudioInputs();
+      this.busy = false;
+    } else {
+      console.warn('getDevices already in progress');
+    }
+  };
+
+  // updateDeviceList is used to handle device enumeration once permissions have been given
+  private updateDeviceList = async () => {
+    const devices: MediaDeviceInfo[] =
+      await navigator.mediaDevices.enumerateDevices();
+    const filtered = devices.filter((device: MediaDeviceInfo) => {
+      return (
+        device.kind === 'audioinput' &&
+        device.deviceId !== '' &&
+        device.label !== ''
+      );
+    });
+    this.devices = filtered;
+  };
+
+  private promptAudioInputs = async () => {
+    const permissions = await getPermissions();
+    if (permissions === 'denied') {
+      this.denied = true;
+      return;
+    }
+
+    // If permissions are prompt, we need to call getUserMedia to ask the user for permission
+    if (permissions === 'prompt') {
+      await this.getUserMedia({
+        audio: true,
+        video: false,
+      });
+    } else {
+      this.updateDeviceList();
+    }
+  };
+}
+const audioDevices = new AudioDevices();
+
+// Here we subscribe to the device state browser event
+// When devices change, the getDevices callback is invoked
+function subscribeDevices(callback) {
+  audioDevices.addEventListener('changeDevices', callback);
+  return () => {
+    audioDevices.removeEventListener('changeDevices', callback);
+  };
+}
+const getDevices = () => audioDevices.devices;
+export function useAudioDevices() {
+  return useSyncExternalStore(subscribeDevices, getDevices, getDevices);
+}
+
+// Here we subscribe to the user's provided permissions
+// When the permission state changes, the useAudioDevices hook is called
+function subscribeDenied(callback) {
+  audioDevices.addEventListener('changeDenied', callback);
+  return () => {
+    audioDevices.removeEventListener('changeDenied', callback);
+  };
+}
+const getDenied = () => audioDevices.denied;
+export function useAudioDenied() {
+  return useSyncExternalStore(subscribeDenied, getDenied, getDenied);
+}
+
+export function useRequestDevices() {
+  return useCallback(() => audioDevices.getDevices(), []);
+}
+
+// getPermissions is used to access the permissions API
+// This API is not fully supported in all browsers so we first check the availability of the API
+async function getPermissions() {
+  if (navigator?.permissions) {
+    return (
+      navigator.permissions
+        // @ts-ignore - ignore because microphone is not in the enum of name for all browsers
+        ?.query({ name: 'microphone' })
+        .then((result) => result.state)
```

Comment:
A similar comment to https://github.com/speechmatics/speechmatics-js-sdk/pull/6#discussion_r1330334594 here

Can we await and wrap with a try catch, as the parent function is async?